rfcs/0006-merkle-tree/0006-merkle-tree.md

```diff
@@ -0,0 +1,137 @@
+---
+Number: 0006
+Category: Standards Track
+Status: Draft
+Author: Ke Wang
+Organization: Nervos Foundation
+Created: 2018-12-01
+---
+
+# Complete Binary Merkle Tree
+
+## Introduction
+
+CKB uses Complete Binary Merkle Tree(CBMT) to generate Merkle Root and Merkle Proof for a static list of items, such as Transactions Root. Basically, CBMT is a complete binary tree, so it has the same node order with the complete binary tree. Compare with other Merkle trees, the hash computation of CBMT is minimal, as well as the proof size.
+
+## Tree Struct
+
+Tree with 6 items([T0, T1, T2, T3, T4, T5]) and tree with 7 items([T0, T1, T2, T3, T4, T5, T6]) is shown below:
+
+```
+        with 6 items                       with 7 items
+
+              B0                                 B0
+             /  \                               /  \
+           /      \                           /      \
+         /          \                       /          \
+       /              \                   /              \
+      B1              B2                 B1              B2
+     /  \            /  \               /  \            /  \
+    /    \          /    \             /    \          /    \
+   /      \        /      \           /      \        /      \  
+  B3      B4      TO      T1         B3      B4      B5      T0
+ /  \    /  \                       /  \    /  \    /  \
+T2  T3  T4  T5                     T1  T2  T3  T4  T5  T6
+```
+
+CBMT can be stored in an array, so the two trees above can be stored as:
+
+```
+[B0, B1, B2, B3, B4, T0, T1, T2, T3, T4, T5]
+[B0, B1, B2, B3, B4, B5, T0, T1, T2, T3, T4, T5, T6]
+```
+
+Suppose we have n items, the size of array would be 2n-1, the index of item i(start at 0) is i+n-1. For node at i, the index of its parent is (i-1)/2, the index of its sibling is (i+1)^1-1 and the indexes of its children are [2i+1, 2i+2].
+
+## Merkle Proof
+
+Merkle Proof can provide a proof for existence of one or more items. Only incalculable nodes that on the path from leaves to root should be included in the proof. If we want to show that [T1, T4] is in the list of 6 items above, only nodes [T5, B3] should be included in the proof.
+
+### Proof Sturct
+
+The schema of proof struct is:
+
+```
+table Proof {
```

Comment:
> The proof struct should include items and items index, in practical applications, the client usually does not know the items' information in advance.

This is just a proof, since in most cases, the client may know some items, so we don't need send the items at first. We should send the items separately if the client need.

Reply:
But the client does not know items index, that means we need extra round to get these info in most cases.

Reply:
> But the client does not know items index, that means we need extra round to get these info in most cases.

you can send them together by a high level struct if it is your case. like:
```
table ItemsWithProof {
    items: [Item];
    proof: Proof;
}
```

Reply:
> But the client does not know items index, that means we need extra round to get these info in most cases.

I think client only needs to send tx hashes to node, and it's easy for node to find out corresponding indices for those hashes. 1 communication route should be enough.

Reply:
> I think client only needs to send tx hashes to node, and it's easy for node to find out corresponding indices for those hashes. 1 communication route should be enough.

Just adding indices to proof is not enough,  the client cannot locate the corresponding index value of hash, that's why I requested to add items hash and index in proof:

```
  items: [[H256, uint32]];
}
```